internal/grpc/grpc.go

```diff
@@ -9,23 +9,49 @@ import (
 	"context"
 	"crypto/tls"
 	"errors"
+	"fmt"
 	"log/slog"
+	"net"
+	"sync"
 
 	grpcRetry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
 
 	"github.com/bufbuild/protovalidate-go"
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/credentials/insecure"
 
+	mpi "github.com/nginx/agent/v3/api/grpc/mpi/v1"
 	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/datasource/host"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/keepalive"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/proto"
 )
 
-// these will come from the agent config
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . GrpcConnectionInterface
+
+type (
+	GrpcConnectionInterface interface {
+		CommandServiceClient() mpi.CommandServiceClient
+		FileServiceClient() mpi.FileServiceClient
+		Close(ctx context.Context) error
+	}
+
+	GrpcConnection struct {
+		config          *config.Config
+		conn            *grpc.ClientConn
+		connectionMutex sync.Mutex
```

Comment:
style: `connectionMutex` is a little verbose if we aren't planning on adding more mutexes to this struct.
https://google.github.io/styleguide/go/decisions.html#external-context-vs-local-names

Reply:
yeah I'll update it there. I probably copy pasted that from somewhere else